foundry.md

```diff
@@ -383,6 +385,96 @@ This rule takes `uint256` value using `#asWord(#range(LM, ARGSTART +Int 4, 32)`
      [priority(40)]
 ```
 
+#### `load` - Loads a storage slot from an address.
+
+```
+function load(address account, bytes32 slot) external returns (bytes32);
+```
+
+`call.load` will match when the `load` function is called at the [Foundry cheatcode address](https://book.getfoundry.sh/cheatcodes/#cheatcodes-reference).
+This rule then loads the storage slot identified by `#asWord(#range(LM, ARGSTART +Int 36, 32))` (referring to `slot` argument) from account `#asWord(#range(LM, ARGSTART +Int 4, 32))` (referring to `account`) and sets the return value to the loaded slot value.
+
+```k
+    rule [call.load]:
+         <k> CALL _ CHEAT_ADDR _VALUE ARGSTART _ARGWIDTH RETSTART RETWIDTH
+          => #loadAccount #asWord(#range(LM, ARGSTART +Int 4, 32))
+          ~> #foundryVmLoad ARGSTART RETSTART RETWIDTH ...
+         </k>
+         <localMem> LM </localMem>
+      requires CHEAT_ADDR ==Int #address(FoundryCheat)
+       andBool #asWord(#range(LM, ARGSTART, 4)) ==Int 1719639408 // selector ( "load(address,bytes32)" )
+      [priority(40)]
+
+
+    syntax KItem ::= "#foundryVmLoad" Int Int Int
+ // ----------------------------------------------------------
+    rule <k> #foundryVmLoad ARGSTART RETSTART RETWIDTH
+          => #setLocalMem RETSTART RETWIDTH #bufStrict(32, #lookup(STORAGE, #asWord(#range(LM, ARGSTART +Int 36, 32))))
+          ~> 1 ~> #push ...
+         </k>
```

Comment:
```suggestion
```k
    rule [call.load]:
         <k> CALL _ CHEAT_ADDR _VALUE ARGSTART _ARGWIDTH RETSTART RETWIDTH
          => #loadAccount #asWord(#range(LM, ARGSTART +Int 4, 32))
          ~> #foundryVmLoad ARGSTART RETSTART RETWIDTH
          ~> 1 ~> #push
          ...
         </k>
         <localMem> LM </localMem>
      requires CHEAT_ADDR ==Int #address(FoundryCheat)
       andBool #asWord(#range(LM, ARGSTART, 4)) ==Int 1719639408 // selector ( "load(address,bytes32)" )
      [priority(40)]


    syntax KItem ::= "#foundryVmLoad" Int Int Int
 // ----------------------------------------------------------
    rule <k> #foundryVmLoad ARGSTART RETSTART RETWIDTH
          => #setLocalMem RETSTART RETWIDTH #bufStrict(32, #lookup(STORAGE, #asWord(#range(LM, ARGSTART +Int 36, 32))))
          ...
         </k>
```

Reply:
It more closely resembles the same control flow of other rules, which means that it may be easier to factor out that control flow later.

Reply:
Yeah, I actually moved `1 ~> #push` around a couple of times. I finally decided to put it inside `#foundryVmLoad` because this is also where we set the return data, so it seemed sensible to let that also be the place to decide whether the call was successful or not (and then either pushing 1 or 0, though in this case `load` cannot fail and we always push 1). Thus, at the moment, the complete semantics of `load` is encapsulated inside `#foundryVmLoad`. Moving `1 ~> #push` to the outer rule would kind of mix the responsibilities of the two rules.

In short, the way the rule is currently written feels "right" to me, but I'm not sure the reason is convincing enough, so let me know :smile:

Reply:
Honestly, I'm not sure we should be creating the account if it doesn't exist, it alters the semantics. Look at how `#accountNonexistent` and `#accountEmpty` are defined, which affect the behavior of the `CALL*` opcodes.

Maybe a test you could write is: "read from the storage slot of some non-existent account, then make a call into that account" (but on a Schedule before `SPURIOUS_DRAGON`, because `Gemptyisnonexistent` is changed on that schedule). So you'd have to set the schedule to some earlier one (`Default`) and some later one (`London')`, and run the tests with both schedules.

I guess we should match what foundry does here. Does it actually create an account, and potentially change the value of `#accountNonexistent`, and `#accountEmpty`? Or does it just return 0 immediately for the storage slot?

Reply:
I followed Foundry here, which creates a new account:
https://github.com/foundry-rs/foundry/blob/014f7c4445dc815e07ae1b469ca31d266bd59fd8/evm/src/executor/inspector/cheatcodes/env.rs#L177-L187

See that they call `load_account()`, which is defined here:
https://github.com/bluealloy/revm/blob/main/crates/revm/src/journaled_state.rs#L478-L504

What I think is still missing from our rules is adding the account to `<accessedAccounts>` and/or `<touchedAccounts>`, which would affect gas usage (I haven't understood the difference between the two these two cells yet). In general, we may want to ensure that our rules use the same amount of gas as in Foundry, though I haven't focused on this yet.

Reply:
I think still the way to test it is to have an example where you load from an account that doesn't exist, then make a call into the account (maybe just an eth transfer), and then check the gas usage afterwards.

We can see if we get the same gas usage as Foundry. It can be a test with no symbolic inputs, and we can assert that the remaining gas is exactly what it should be.

Reply:
If we don't update `touched` and `accessed` correctly, then the gas usage will be different. Having this test in place will tell us whether we are conformant to Foundry (when it changes too).

Reply:
`touched` does not seem to be used by the semantics. It is only ever read in `evm-node.md`, but I'm not sure for what purpose and which addresses should be considered part of this set